File: crates/ms-mde/tests/enrollment.rs
// use pretty_assertions::assert_eq;

// #[test]
// fn test_enrollment_request() {
//     let payload = r#" <s:Envelope xmlns:a="http://www.w3.org/2005/08/addressing"
//     xmlns:s="http://www.w3.org/2003/05/soap-envelope">
//       <s:Header>
//          <a:Action s:mustUnderstand="1">
//             http://schemas.microsoft.com/windows/pki/2009/01/enrollment/RST/wstep
//          </a:Action>
//          <a:MessageID>urn:uuid:b5d1a601-5091-4a7d-b34b-5204c18b5919</a:MessageID>
//          <a:ReplyTo>
//             <a:Address>http://www.w3.org/2005/08/addressing/anonymous</a:Address>
//          </a:ReplyTo>
//          <wsse:Security>
//             <wsse:BinarySecurityToken
//                ValueType=http://schemas.microsoft.com/5.0.0.0/
//     ConfigurationManager/Enrollment/DeviceEnrollmentUserToken
//               EncodingType=""http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd#base64binary"">
//                <!-- Token Removed -->
//             </wsse:BinarySecurityToken>
//          </wsse:Security>
//       </s:Header>
//       <s:Body xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
//     xmlns:xsd="http://www.w3.org/2001/XMLSchema">
//          <wst:RequestSecurityToken xmlns="http://docs.oasis-open.org/ws-sx/ws-trust/200512">
//             <wst:TokenType>
//     "http://schemas.microsoft.com/5.0.0.0/ConfigurationManager/Enrollment/DeviceEnrollmentToken"
//             </wst:TokenType>
//             <wst:RequestType>
//                http://docs.oasis-open.org/ws-sx/ws-trust/200512/Issue
//             </wst:RequestType>
//             <wsse:BinarySecurityToken
//                EncodingType="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd#base64binary"
//                ValueType="http://schemas.microsoft.com/windows/pki/2009/01/enrollment#PKCS10"
//                xmlns="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd">
//     <!-- Token removed -->
//             </wsse:BinarySecurityToken>
//             <ac:AdditionalContext>
//                <ac:ContextItem Name="DeviceType">
//                   <ac:Value>CIMClient_Windows</ac:Value>
//                </ac:ContextItem>
//             </ac:AdditionalContext>
//           </wst:RequestSecurityToken>
//        </s:Body>
//     </s:Envelope>"#;

//     todo!();
// }

// #[test]
// fn test_enrollment_response() {
//     let payload = r#"<s:Envelope
//     xmlns:s="http://www.w3.org/2003/05/soap-envelope"
//     xmlns:a="http://www.w3.org/2005/08/addressing">
//     <s:Header>
//       <a:Action s:mustUnderstand="1">
//          http://schemas.microsoft.com/windows/pki/2009/01/enrollment/RSTRC/wstep
//       </a:Action>
//     </s:Header>
//     <s:Body>
//        <wst:RequestSecurityTokenResponseCollection
//           xmlns="http://docs.oasis-open.org/ws-sx/ws-trust/200512">
//           <wst:RequestSecurityTokenResponse>
//              <wst:RequestedSecurityToken>
//                 <wst:TokenType>
//  "http://schemas.microsoft.com/5.0.0.0/ConfigurationManager/Enrollment/DeviceEnrollmentToken"
//                 </wst:TokenType>
//                 <wsse:BinarySecurityToken
//                    ValueType="http://schemas.microsoft.com/5.0.0.0/Configuration Manager/Enrollment/DeviceEnrollmentProvisionDoc"
//                    EncodingType="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd#base64binary">

//                    <!-- base64 encoded provisioning document removed -->
//                 </wsse:BinarySecurityToken>
//              </wst:RequestedSecurityToken>
//           </wst:RequestSecurityTokenResponse>
//        </wst:RequestSecurityTokenResponseCollection>
//     </s:Body>
//  </s:Envelope>"#;

//     todo!();
// }

File: packages/ms-mde/src/enrollment.ts
import * as v from "valibot";
import { envelope, withEnvelope } from "./envelope";
import { serializeXml } from "./util";
import type { WapProvisioningProfile } from "./wap";

// All of the known context items with a fallback to any generic ones.
// *Be aware* this can't be treated as an exhaustive list as the MDM client may send additional context items or omit ones listed here.
// TODO: Are we gonna need this? Should it be a schema for validation?
export type ContextItem =
	| {
			"ac:Value": boolean;
			"@_Name": "UXInitiated" | "TargetedUserLoggedIn" | "NotInOobe";
	  }
	| {
			"ac:Value": string;
			"@_Name":
				| "HWDevID"
				| "Locale"
				| "DeviceName"
				| "MAC"
				| "DeviceID"
				| "EnrollmentType"
				| "DeviceType"
				| "OSVersion"
				| "ApplicationVersion"
				| "Manufacturer"
				| "DeviceModel"
				| "SerialNumber";
	  }
	| {
			"ac:Value": number;
			"@_Name": "OSEdition" | "RequestVersion";
	  }
	| {
			"ac:Value": string;
			"@_Name": string;
	  };

//  contains the device information and identity certificate CSR
export const enrollmentRequest = envelope(
	"http://schemas.microsoft.com/windows/pki/2009/01/enrollment/RST/wstep",
	{
		"wst:RequestSecurityToken": v.object({
			"wst:TokenType": v.literal(
				"http://schemas.microsoft.com/5.0.0.0/ConfigurationManager/Enrollment/DeviceEnrollmentToken",
			),
			"wst:RequestType": v.literal(
				"http://docs.oasis-open.org/ws-sx/ws-trust/200512/Issue",
			),
			"wsse:BinarySecurityToken": v.object({
				"#text": v.string(),
				"@_ValueType": v.literal(
					"http://schemas.microsoft.com/windows/pki/2009/01/enrollment#PKCS10",
				),
				"@_EncodingType": v.literal(
					"http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd#base64binary",
				),
			}),
			"ac:AdditionalContext": v.object({
				"@_xmlns": v.literal(
					"http://schemas.xmlsoap.org/ws/2006/12/authorization",
				),
				"ac:ContextItem": v.array(
					v.object({
						"ac:Value": v.union([v.string(), v.number(), v.boolean()]),
						"@_Name": v.string(),
					}),
				),
			}),
		}),
	},
	{
		schemas: {
			u: "http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd",
			wsse: "http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd",
			wst: "http://docs.oasis-open.org/ws-sx/ws-trust/200512",
			ac: "http://schemas.xmlsoap.org/ws/2006/12/authorization",
		},
	},
);

// contains the management client configuration and signed identity certificate
export const enrollmentResponse = withEnvelope(
	"http://schemas.microsoft.com/windows/pki/2009/01/enrollment/RSTRC/wstep",
	(opts: {
		wapProvisioningProfile: object;
	}) => ({
		RequestSecurityTokenResponseCollection: {
			"@_xmlns": "http://docs.oasis-open.org/ws-sx/ws-trust/200512",
			RequestSecurityTokenResponse: {
				TokenType:
					"http://schemas.microsoft.com/5.0.0.0/ConfigurationManager/Enrollment/DeviceEnrollmentToken",
				DispositionMessage: {
					"@_xmlns":
						"http://schemas.microsoft.com/windows/pki/2009/01/enrollment",
				},
				RequestedSecurityToken: {
					BinarySecurityToken: {
						"@_xmlns":
							"http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd",
						"@_ValueType":
							"http://schemas.microsoft.com/5.0.0.0/ConfigurationManager/Enrollment/DeviceEnrollmentProvisionDoc",
						"@_EncodingType":
							"http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd#base64binary",
						"#text": btoa(
							serializeXml(opts.wapProvisioningProfile)
								.replaceAll("\n", "")
								.replaceAll("\t", ""),
						),
					},
				},
				RequestID: {
					"@_xmlns":
						"http://schemas.microsoft.com/windows/pki/2009/01/enrollment",
					"#text": "0",
				},
			},
		},
	}),
	{
		schemas: {
			u: "http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd",
		},
		"s:Header": {
			"o:Security": {
				"@_xmlns:o":
					"http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd",
				"@_s:mustUnderstand": "1",
				"u:Timestamp": {
					"@_u:Id": "_0",
					"u:Created": "2018-11-30T00:32:59.420Z", // TODO
					"u:Expires": "2018-12-30T00:37:59.420Z", // TODO
				},
			},
		},
	},
);
